GitUI/UserControls/FileStatusList.cs

```diff
@@ -697,64 +703,146 @@ public void SetDiffs(IReadOnlyList<GitRevision> revisions, Func<ObjectId, GitRev
+            {
+                // No variant of range diff, show multi diff
+                tuples.AddRange(
+                    revisions
+                        .Where(rev => rev != firstRev && rev != selectedRev)
+                        .Select(rev => new FileStatusWithDescription
+                        {
+                            firstRev = rev,
+                            secondRev = selectedRev,
+                            summary = _diffWithParent.Text + GetDescriptionForRevision(rev.ObjectId),
+                            statuses = Module.GetDiffFilesWithSubmodulesStatus(rev.ObjectId, selectedRev.ObjectId, selectedRev.FirstParentId)
+                        }));
+
+                GitItemStatusesWithDescription = tuples;
+                return;
+            }
+
+            // Present common files in BASE->B, BASE->A separately
+            // For the following diff:  A->B a,c,d; BASE->B a,b,c; BASE->A a,b,d
+            // (the file a has unique changes, b has the same change and c,d is changed in one of the branches)
+            // The following groups will be shown: A->B a,c,d; BASE->B a,c; BASE->A a,d; Common BASE b
+            var allBaseToB = Module.GetDiffFilesWithSubmodulesStatus(baseRevGuid, selectedRev.ObjectId, selectedRev.FirstParentId);
+            var allBaseToA = Module.GetDiffFilesWithSubmodulesStatus(baseRevGuid, firstRev.ObjectId, firstRev.FirstParentId);
+
+            var comparer = new GitItemStatusNameEqualityComparer();
+            var commonBaseToAandB = allBaseToB.Intersect(allBaseToA, comparer).Except(allAToB, comparer).ToList();
+
+            var revBase = new GitRevision(baseRevGuid);
+            tuples.Add(new FileStatusWithDescription
+            {
+                firstRev = revBase,
+                secondRev = selectedRev,
+                summary = _diffBaseToB.Text + GetDescriptionForRevision(selectedRev.ObjectId),
+                statuses = allBaseToB.Except(commonBaseToAandB, comparer).ToList()
+            });
+            tuples.Add(new FileStatusWithDescription
+            {
+                firstRev = revBase,
+                secondRev = firstRev,
+                summary = _diffBaseToB.Text + GetDescriptionForRevision(firstRev.ObjectId),
+                statuses = allBaseToA.Except(commonBaseToAandB, comparer).ToList()
+            });
+            tuples.Add(new FileStatusWithDescription
+            {
+                firstRev = revBase,
+                secondRev = selectedRev,
+                summary = _diffCommonBase.Text + GetDescriptionForRevision(baseRevGuid),
+                statuses = commonBaseToAandB
+            });
+
+            // Git range-diff is slow and memory consuming, so just skip if diff is large
+            // to avoid that GE seem to hang when selecting the range diff
+            const int maxRangeDiffCommits = 100;
+            int? count = Module.GetCommitCount(firstRevHead.ToString(), selectedRevHead.ToString());
```

Comment:
This seems to work. I am confused how it does - at least without digging into git commands:
For `baseA..headA baseB..headB`, this will try to find the number of commits between the _unrelated_ `headA` and `headB`.
Would one or both distances between `baseA` and `headA` as well as between `baseB` and `headB` be a more comprehensible criterion?
(I have not yet thought about which one must be chosen for a selection of 3 commits.)

Reply:
baseA/baseB is only used with four selections, there is a check that baseA really is the base to firstSelected, otherwise multi diff is done. So this is already checked
For three commits, the merge-base must be selected (so a 'base' that is older is not identified and multidiff is displayed instead).

Reply:
Yes, `headA` and `headB` have a common ancestor here. But still, what does "commit count" between these two mean? They are unrelated -- none is the base of the other -- at least in the interesting use case.

Reply:
The commit count are all commits traversing from first/A to selected/B. Compare to a local branch ahead 2 and behind 3 the tracked branch would have a commit count of 5.
It seems like git-range-diff is o(2) or so; while 100 seem fine for me, 1000 or so was slow.

Reply:
The git command used is "git rev-list headA ^headB --count". It seems to not take much time. And it yields smaller counts than you expect.

Since the range-diff is not run anymore, unless it is selected: Do we still need to suppress it if there are too many commits inbetween?

If yes, I think my initial remarks should be taken into account:

> For `baseA..headA baseB..headB`, this will try to find the number of commits between the _unrelated_ `headA` and `headB`.
Would one or both distances between `baseA` and `headA` as well as between `baseB` and `headB` be a more comprehensible criterion?

Reply:
The command to find differences should be 
```git rev-list  --count headA...headB```
will update

There is still possible to have larger differences for commands with diverging forks when range-diff can take some time

At work with the Symantec virus running, range-diff requires 20-40s for a couple of range-diff with a count of about 110 (time is quicker if many commits are the same)
Not responsive, but you select the range-diff explicitly and it does not 'hang' forever
These are scenarios where I am interested to see the result (with this many, only the overview i)

For two forks with a count of 500 I waited 20 min before aborting
The range-diff algorithm has ```cubic runtime complexity```

I find 100 reasonable
I could be interested with more, if  implement some async calculations that can be aborted

This leads into another presentation change: I would like to make it possible to only see the summary.
(Parsing the output yield that, but adds a delay.) Will probably add to the presentation part of the PR for that.

Reply:
One way to set the limit quite low is to include the Git command as text, then it is relative easy for the user to get the output without risking "hanging".

Reply:
Changed commit count and added the following when exceeding

![image](https://user-images.githubusercontent.com/6248932/95517262-f4c73980-09c0-11eb-9ea7-aa7acad8ab79.png)

Reply:
Will add one more tweak: Before running the range-diff command, show the command that runs
![image](https://user-images.githubusercontent.com/6248932/95613958-07477e80-0a66-11eb-9223-3046fe938f84.png)

If the command is very slow, you at least see what is going on